### PR TITLE
Layers: Implement unique_objects layer using vk.xml

### PR DESCRIPTION
**Pull request by mark-lunarg**
```
UniqueObjects was the final LVL layer not using vk.xml for codegen.
The new implementation follows the threading layer model which
(outside of safe_struct dependencies) is more tolerant of API
changes. Also updated for structure and style.

build-android/android-generate.bat - Updated for build
build-android/android-generate.sh  - Updated for build
generator.py                       - Added Unique Objects Generator
genvk.py                           - Added call to U_O generator
layers/CMakeLists.txt              - Updated for build
layers/unique_objects.cpp          - New: manually generated code
layers/unique_objects.h            - Now a true header file
vk-layer-generate.py               - killed!
```

### CODE REVIEW

File: layers/unique_objects.cpp [hunk 1]
```diff
@@ -0,0 +1,756 @@
...
+    std::unique_ptr<safe_VkMemoryAllocateInfo> safe_allocate_info;
+    std::unique_ptr<safe_VkDedicatedAllocationMemoryAllocateInfoNV> safe_dedicated_allocate_info;
+    layer_data *device_data = get_my_data_ptr(get_dispatch_key(device), layer_data_map);
+
+    if ((pAllocateInfo != nullptr) &&
+        ContainsExtStruct(pAllocateInfo, VK_STRUCTURE_TYPE_DEDICATED_ALLOCATION_MEMORY_ALLOCATE_INFO_NV)) {
+        // Assuming there is only one extension struct of this type in the list for now
+        safe_dedicated_allocate_info =
+            std::unique_ptr<safe_VkDedicatedAllocationMemoryAllocateInfoNV>(new safe_VkDedicatedAllocationMemoryAllocateInfoNV);
+        safe_allocate_info = std::unique_ptr<safe_VkMemoryAllocateInfo>(new safe_VkMemoryAllocateInfo);
```
Comment (tobine) [inline layers/unique_objects.cpp]: Could pass pAllocateInfo to the constructor here and avoid the initialize call two lines below

File: layers/unique_objects.cpp [hunk 2]
```diff
@@ -0,0 +1,756 @@
...
+        return NULL;
+    }
+    return disp_table->GetInstanceProcAddr(instance, funcName);
+}
+
+VKAPI_ATTR VkResult VKAPI_CALL AllocateMemory(VkDevice device, const VkMemoryAllocateInfo *pAllocateInfo,
+                                              const VkAllocationCallbacks *pAllocator, VkDeviceMemory *pMemory) {
+    const VkMemoryAllocateInfo *input_allocate_info = pAllocateInfo;
+    std::unique_ptr<safe_VkMemoryAllocateInfo> safe_allocate_info;
+    std::unique_ptr<safe_VkDedicatedAllocationMemoryAllocateInfoNV> safe_dedicated_allocate_info;
```
Comment (tobine) [inline layers/unique_objects.cpp]: Could declare this in the "if" block below

File: layers/unique_objects.cpp [hunk 3]
```diff
@@ -0,0 +1,756 @@
...
+#include <list>
+#include <memory>
+
+#include "vk_loader_platform.h"
+#include "vulkan/vk_layer.h"
+#include "vk_layer_config.h"
+#include "vk_layer_extension_utils.h"
+#include "vk_layer_utils.h"
+#include "vk_enum_validate_helper.h"
+#include "vk_struct_validate_helper.h"
```
Comment (mark-lunarg) [inline layers/unique_objects.cpp]: Done.

File: layers/unique_objects.cpp (hunk 2 above)
Comment (mark-lunarg) [inline layers/unique_objects.cpp]: Done.

**Comment by mark-lunarg** [comment]
Updated and pushed to Master.

**Pull request merged**